### PR TITLE
Fix(payment): STRIPE-53 Give more information on payment authentication with 3ds failure error message

### PR DESCRIPTION
**Pull request by JavvmBC**
## What? [STRIPE-53](https://bigcommercecloud.atlassian.net/browse/STRIPE-53)
Alteration of error message when stripe 3DS auth fails

## Why?
To give more information to the user about the failure

## Testing / Proof
Screen Shot 2022-07-04 at 11 38 23

@bigcommerce/checkout @bigcommerce/payments @bigcommerce/apex-integrations 

## Depends on 
https://github.com/bigcommerce/checkout-js/pull/940

### CODE REVIEW

File: packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 1]
```diff
@@ -485,7 +485,12 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
                 if (this._isCancellationError(result.error)) {
                     throw new PaymentMethodCancelledError();
                 }
-                throw new Error(result.error.message);
+
+                if (result.error.code === 'payment_intent_authentication_failure') {
+                  throw new Error('User did not authenticate');
+                } else {
+                    throw new Error(result.error.message);
+                }
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.ts:493]: For this change, please refer to #1286 and #1367 and their sibling PRs in order to translate your custom message.

```suggestion
                if (this._isAuthError()) {
                    throw new StripeV3Error(StripeV3ErrorType.AuthFailure);
                }

                throw new Error(result.error.message);
```

**Review by mauricio-sg — CHANGES_REQUESTED** [review]
You'll need some additional changes.

File: packages/core/src/payment/strategies/stripev3/stripev3-error.spec.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import StripeV3Error, { StripeV3ErrorType } from './stripev3-error';
+
+describe('StripeV3Error', () => {
+    it('returns error name, type and message', () => {
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-error.spec.ts:4]: The test description doesn't match the assertions.

File: packages/core/src/payment/strategies/stripev3/stripev3-error.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+const defaultMessage = 'There was an error while processing your payment. Please try again or contact us.';
+
+export enum StripeV3ErrorType {
+    AuthFailure = 'auth_failure',
+}
+
+export default class StripeV3Error extends StandardError {
+    subtype: string;
+    constructor(subtype: StripeV3ErrorType) {
+        super(getErrorMessage(subtype) || defaultMessage);
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-error.ts:12]: What's the purpose of `defaultMessage`? It's unreachable code so far.

File: packages/core/src/payment/strategies/stripev3/stripev3-error.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { StandardError } from '../../../common/error/errors';
+
+const defaultMessage = 'There was an error while processing your payment. Please try again or contact us.';
+
+export enum StripeV3ErrorType {
+    AuthFailure = 'auth_failure',
+}
+
+export default class StripeV3Error extends StandardError {
+    subtype: string;
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-error.ts:10]: Please add a new line.

File: packages/core/src/payment/strategies/stripev3/stripev3-initialize-options.ts [hunk 5]
```diff
@@ -59,4 +61,9 @@ export default interface StripeV3PaymentInitializeOptions {
      * Hosted Form Validation Options
      */
     form?: HostedFormOptions;
+
+    /**
+     * 3ds Authentication Error
+     */
+    stripev3AuthError?: Stripev3Error;
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-initialize-options.ts:68]: I don't see this property being used anywhere. 🤔

File: packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts [hunk 6]
```diff
@@ -775,6 +775,35 @@ describe('StripeV3PaymentStrategy', () => {
...
+                        three_ds_result: {
+                            token: 'token',
+                        },
+                    }));
+                    const authenticationFailErrorResponse = new RequestError(getResponse({
+                        ...getErrorPaymentResponseBody(),
+                        errors: [
+                            { code: 'payment_intent_authentication_failure' },
+                        ],
+                    }));
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts:793]: I think this is not necessary. 🤔

File: packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts [hunk 7]
```diff
@@ -775,6 +775,35 @@ describe('StripeV3PaymentStrategy', () => {
...
+                    const stripeErrorMessage = 'Stripe error message.';
+
+                    jest.spyOn(paymentActionCreator, 'submitPayment')
+                        .mockReturnValueOnce(of(createErrorAction(PaymentActionType.SubmitPaymentFailed, threeDSecureRequiredErrorResponse)))
+                        .mockReturnValueOnce(of(createErrorAction(PaymentActionType.SubmitPaymentFailed, authenticationFailErrorResponse)));
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts:798]: I think this is not necessary. 🤔

**Review by mauricio-sg — CHANGES_REQUESTED** [review]

File: packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts (hunk 7 above)
Comment (JavvmBC) [reply packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts]: Thank you for the comments, already corrected them and commit them to this PR. Let me know if everything is correct 😎

File: packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts [hunk 8]
```diff
@@ -775,6 +775,29 @@ describe('StripeV3PaymentStrategy', () => {
...
+                        ...getErrorPaymentResponseBody(),
+                        errors: [
+                            { code: 'three_d_secure_required' },
+                        ],
+                        three_ds_result: {
+                            token: 'token',
+                        },
+                    }));
+
+                    const stripeErrorMessage = 'Stripe error message.';
```
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts:789]: I think this const is unnecessary.

**Review by mauricio-sg — APPROVED** [review]
I have two comments but not a blocker. LGTM.

**Comment by JavvmBC** [comment]
Thank you for the comments 🦸

File: packages/core/src/payment/strategies/stripev3/stripev3-error.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { StandardError } from '../../../common/error/errors';
+
+export enum StripeV3ErrorType {
+    AuthFailure = 'auth_failure',
+}
+
+export default class StripeV3Error extends StandardError {
```
Comment (animesh1987) [inline packages/core/src/payment/strategies/stripev3/stripev3-error.ts:7]: Do we need to add a new error class for this case? Can we leverage a previous one?

Comment (JavvmBC) [reply packages/core/src/payment/strategies/stripev3/stripev3-error.ts:7]: I created the class so we can facilite for future the handling of stripe v3

**Review by animesh1987 — APPROVED** [review]

**Pull request merged**